### ISSUE TITLE
assert_eq: Better error messages

### ISSUE THREAD

**Issue by sharkdp** [sharkdp/numbat#505]
Consider this example:
```
>>> assert_eq(2m, 201cm, 0.1cm)
error: Assertion failed
  ┌─ <input:2>:1:15
  │
1 │ assert_eq(2m, 201cm, 0.1cm)
  │           --  ^^^^^ 201 cm
  │           │    
  │           2 m
  │
  = Assertion failed because the following two quantities differ by more than 0.1 cm:
      2 m
      201 cm
```
This is not very helpful, since we show the LHS/RHS values in different units.

And if the contrived example is not enough, here's a real world error message that I got today:
```
>>> assert_eq(-77° + 0′ + 32″, -77.0089°, 1e-4°)
error: Assertion failed
  ┌─ <input:1>:1:28
  │
1 │ assert_eq(-77° + 0′ + 32″, -77.0089°, 1e-4°)
  │           ---------------  ^^^^^^^^^ -77.0089°
  │           │                 
  │           -1.34375
  │
  = Assertion failed because the following two quantities differ by more than 0.00000174533:
      -1.34375
      -77.0089°
```

We certainly want to do a unit conversion here (to the same unit that is also used for the actual comparison, I would guess the LHS unit). Maybe we even want to show a comparison in the two possible units (the LHS unit and the RHS unit)?

**Comment by xmbhasin** [comment]
I'm thinking of contributing this feature.

I think converting both lhs and rhs to epsilon (the maximum expected difference) would be the most intuitive and would allow a user to see the difference in different units easily if desired by changing the unit of epsilon and without having to mess with the lhs or rhs expressions.

For example:

```
>>> assert_eq(2m, 201cm, 0.1cm)
error: Assertion failed
  ┌─ <input:1>:1:15
  │
1 │ assert_eq(2m, 201cm, 0.1cm)
  │           --  ^^^^^ 201 cm (201 cm)
  │           │
  │           200 cm (2 m)
  │
  = Assertion failed because the following two quantities differ by more than 0.1 cm:
      200 cm (2 m)
      201 cm (201 cm)
  ```
  
  ```
  >>> assert_eq(2m, 201cm, 0.1mm)
error: Assertion failed
  ┌─ <input:4>:1:15
  │
1 │ assert_eq(2m, 201cm, 0.1mm)
  │           --  ^^^^^ 2010.0 mm (201 cm)
  │           │
  │           2000 mm (2 m)
  │
  = Assertion failed because the following two quantities differ by more than 0.1 mm:
      2000 mm (2 m)
      2010.0 mm (201 cm)
  ```
  
The original values are written in parentheses but it feels a bit redundant, so I could be convinced to remove it.

Note that, in the second example with degrees, epsilon is converted to radians at some point before the core assertion is handled. So with this change the comparison would be done in radians. Expressions like `-77° + 0′ + 32″` are also converted to radians automatically.

```
>>> assert_eq(-77° + 0′ + 32″, -78.0089°, 1°)
error: Assertion failed
  ┌─ <input:6>:1:28
  │
1 │ assert_eq(-77° + 0′ + 32″, -78.0089°, 1°)
  │           ---------------  ^^^^^^^^^ -1.36151 (-78.0089°)
  │           │
  │           -1.34375 (-1.34375)
  │
  = Assertion failed because the following two quantities differ by more than 0.0174533:
      -1.34375 (-1.34375)
      -1.36151 (-78.0089°)
```

**Comment by sharkdp** [comment]
> I'm thinking of contributing this feature.

Cool! Let us know if you need help.

> I think converting both lhs and rhs to epsilon (the maximum expected difference) would be the most intuitive and would allow a user to see the difference in different units easily if desired by changing the unit of epsilon and without having to mess with the lhs or rhs expressions.

This sounds very reasonable, and your example output looks great to me.

> The original values are written in parentheses but it feels a bit redundant, so I could be convinced to remove it.

I like it.



> Note that, in the second example with degrees, epsilon is converted to radians at some point before the core assertion is handled.

Oh, ok. This is something very specific to angles, since they "decay" to scalars (in radian), as you observed. So let's ignore this for now and maybe discuss this elsewhere. If you are interested in the topic, see also: #167

**Comment by xmbhasin** [comment]
One thing that I think is missing after #527, is allowing users to get more precision in the error message by adding 0s to epsilon. 

In the example below, ideally I think it would show 3 decimal places of precision:
```
assert_eq(212121000.123456cm, 212121001.123456cm, 0.900cm)
error: Assertion failed
  ┌─ <input:23>:1:31
  │
1 │ assert_eq(212121000.123456cm, 212121001.123456cm, 0.900cm)
  │           ------------------  ^^^^^^^^^^^^^^^^^^ 212121001.1 cm
  │           │
  │           212121000.1 cm
  │
  = Assertion failed because the following two quantities differ by 1.0 cm, which is more than 0.9 cm:
      212121000.1 cm
      212121001.1 cm
```

Implementing this would probably require some way to carry the user's desired precision alongside the epsilon Quantity.

To work around this one would need to add a 1 in the last decimal place, though this could change the comparison slightly.

**Comment by sharkdp** [comment]
I'm not super concerned about the trailing zero problem.

But I found some cases where the current implementation leads to problems. It might only be relevant for angles since they have this weird simplification behavior. Take a look at the new output for the input suggested in the first post:

```
>>> assert_eq(-77° + 0′ + 32″, -77.0089°, 1e-4°)
error: Assertion failed
  ┌─ <input:2>:1:28
  │
1 │ assert_eq(-77° + 0′ + 32″, -77.0089°, 1e-4°)
  │           ---------------  ^^^^^^^^^ -1.3440588583390611000000 (-77.0089°)
  │           │                 
  │           -1.3437483836576787000000
  │
  = Assertion failed because the following two quantities differ by 0.0003104746813824555000, which is more than 0.0000017453292519943296:
      -1.3437483836576787000000
      -1.3440588583390611000000 (-77.0089°)
```

That illustrates a few problems:
- Epsilon is converted to a scalar at some point so we compare scalar values
- In the deg->rad conversion, `1e-4°` turns into a irrational number, which means that we use the maximum number of digits. Far too many for the given precision.
- We don't show the value in degrees for the left-hand-side argument.

**Comment by sharkdp** [comment]
Oh hold on. That will actually be fixed by #537. I'll add a regression test in there.
